src/librustc/hir/check_attr.rs

```diff
@@ -16,6 +16,8 @@ use std::fmt::{self, Display};
 use syntax::{attr, symbol::sym};
 use syntax_pos::Span;
 
+use rustc_error_codes::*;
```

Comment:
Hmm; why do this in each of the files? Shouldn't we change the diagnostics macros to not use constants at all but rather to just do the stringification? Tidy can do the validation later to make sure that there are no unused error codes or references to codes not registered. This way we don't have to change any of the `Cargo.toml` files here except for librustc_interface.

Reply:
I don't agree with this: it allows us to be sure that the used error codes exist at compile time. If we switch to strings, we won't be able to have this check anymore (or at least, not at compile time).

Reply:
I'm confused... In the issue you noted that doing what you are doing here would make the recompilation issue worse, so what changed? -- Tidy is part of "compile time" for the purposes of CI so moving the checks there still means problems are rejected. Moreover, eventually and ideally we want to make this crate only have rustc_driver as the only reverse dependency and maybe even link this in dynamically so that it can be swapped for something else at runtime (e.g. for Spanish translations or whatnot).

Reply:
The problem is that, if we do it like this, the only way to check that a used error code exists will be to add another check after the compiler is built. I prefer re-building some dependencies rather than having to wait the whole build to see if I didn't forget to create a new error code.

Reply:
If that's an issue then we can move the tidy checks (or this specific check) early or something and always execute them on e.g. `./x.py test -i src/test/ui --stage 1 --pass check` which is the common command to do. We could also execute tidy on `./x.py check`. cc @Mark-Simulacrum 

The way you've implemented this now, if I were to e.g. add a new error code I would need to rebuild pretty much everything, including `librustc`, instead of just rebuilding `librustc_interface`, which goes quickly. This would be painful if all you want to do is to `--bless` some UI tests.

Reply:
Ok, then I suggest this: let's open an issue about this once this PR is merged and handle it into another PR. I think this one is already \*huge\* enough.

Reply:
Fair enough. :)

Reply:
I am against adding anything to x.py check or x.py test that runs by default and isn't tidy. You can always run tidy first if you want.

Reply:
@Mark-Simulacrum What about always running tidy on said commands? -- I personally have been bitten many times by forgetting to do so since I typically never use the full `./x.py test` and I always forget to run tidy manually.

Reply:
I am against running tidy implicitly unless we can make it much faster; currently it's 1-3 seconds of overhead which is way to much for a "always run" type command.